stan/math/fwd/fun/accumulator.hpp

```diff
@@ -0,0 +1,85 @@
+#ifndef STAN_MATH_FWD_FUN_ACCUMULATOR_HPP
+#define STAN_MATH_FWD_FUN_ACCUMULATOR_HPP
+
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/fwd/core.hpp>
+#include <stan/math/fwd/meta.hpp>
+#include <stan/math/fwd/fun/sum.hpp>
+#include <stan/math/prim/fun/accumulator.hpp>
+#include <vector>
+#include <type_traits>
+
+namespace stan {
+namespace math {
+
+/**
+ * Class to accumulate values and eventually return their sum.  If
+ * no values are ever added, the return value is 0.
+ *
+ * This class is useful for speeding up autodiff of long sums
+ * because it uses the <code>sum()</code> operation (either from
+ * <code>stan::math</code> or one defined by argument-dependent lookup.
+ *
+ * @tparam T Type of scalar added
+ */
+template <typename T>
+class accumulator<fvar<T>> {
```

Comment:
I still think we dont need a fwd specialization for accumulator. At worst we might need a `.hpp` file for it that just includes `/fwd/fun/sum.hpp` and `prim/fun/accumulator.hpp`

Reply:
So I tried removing this and it failed to compile. This is definitely an include issue but a pretty big one that I think should be tackled in a separate PR that's going to take some time to sort out. I'm going to make an issue about it today or Sunday

Reply:
I would agree with this reasoning if the issue was present in develop before this PR. But this PR introduces the issue of duplicated code, so I dont think we should merge it until the issue is resolved.

I am fine with resolving includes in a separate PR, but I would wait with merging of this PR until the includes are resolved.

Reply:
So I'm pretty sure this issue actually did exist before this PR but we didn't notice. The fundamental problem is in the godbolt below

https://godbolt.org/z/je89G573e

Before the class definition of `accumulate` we need to have the definition of `sum()` in order to compile with the `var` specialized sum. It was working before because one of the definitions of `sum()` in prim is 

```cpp
template <typename T>
inline T sum(const std::vector<T>& m) {
  return std::accumulate(m.begin(), m.end(), T{0});
}
```

Which will work for `var`, `double`, `fvar`, etc. We can have the include order anyway we want when we are just including functions, but when we include classes we need to make sure that the definitions that the class uses are available before the class is declared. We don't see this error very often because most things in Stan math are just functions (though this error did just pop up again with `apply_scalar_unary` where the definition for `apply_scalar_unary<var>` did not show up before prim's `log()` function.

I can take a shot at fixing it today, but it's going to require some fancy footwork and a good headache to make sure we get the includes correct. I'm worries this is going to take me a while to fix and this is stopping the new matrix stuff from working in the compiler so if your okay with it I'd prefer merging this PR in and then tacking this stuff in a seperate issue.

Reply:
We did have multiple issues with include order before, but we never duplicated classes just because of that. So far these  issues could be solved by changing the order of includes. 

Duplicating the code is a new problem introduced in this PR and I would like it resolved before this is merged. 

Maybe I will be able to take some time to look at it today.

Reply:
^ if you can I'd totally appreciate it. I took another stab at the includes today and couldn't find an answer that wasn't funky.